# SQL анализ базы данных книг

## Задача:
Проанализировать базу данных с информацией о книгах, издательствах, авторах, а также обзоры книг.

## Цель: 
Сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args = {'sslmode':'require'}) 

In [2]:
print('books - Содержит данные о книгах:')
query = '''SELECT *
           FROM books'''
books = pd.io.sql.read_sql(query, con = engine) 
display(books.head())
#display(books.info())

books - Содержит данные о книгах:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
print('authors - Содержит данные об авторах:')
query = '''SELECT *
           FROM authors'''
authors = pd.io.sql.read_sql(query, con = engine) 
display(authors.head())
#display(authors.info())

authors - Содержит данные об авторах:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
print('publishers - Содержит данные об издательствах:')
query = '''SELECT *
           FROM publishers'''
publishers = pd.io.sql.read_sql(query, con = engine) 
display(publishers.head())
#display(publishers.info())

publishers - Содержит данные об издательствах:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
print('ratings - Содержит данные о пользовательских оценках книг:')
query = '''SELECT *
           FROM ratings'''
ratings = pd.io.sql.read_sql(query, con = engine) 
display(ratings.head())
#display(ratings.info())

ratings - Содержит данные о пользовательских оценках книг:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
print('reviews - Содержит данные о пользовательских обзорах на книги:')
query = '''SELECT *
           FROM reviews'''
reviews = pd.io.sql.read_sql(query, con = engine) 
display(reviews.head())
display(reviews.info())

reviews - Содержит данные о пользовательских обзорах на книги:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера
Отлично
    
Доступ к базе получили, на таблицы посмотрели, переходим к основной работе
    
</div>

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [7]:
query = ''' SELECT COUNT(book_id)
            FROM books
            WHERE CAST(publication_date AS DATE) > '2000-01-01' '''

In [8]:

pd.io.sql.read_sql(query, con = engine)

count
0    819

**819** книг вышло после 1 января 2000 года

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [9]:
query = ''' SELECT b.title,
                   COUNT(DISTINCT rev.review_id),
                   ROUND(AVG(rat.rating), 1)
            FROM ratings AS rat
            LEFT JOIN reviews AS rev ON rev.book_id = rat.book_id 
            LEFT JOIN books AS b ON b.book_id = rat.book_id 
            GROUP BY b.book_id
            ORDER BY COUNT(DISTINCT rev.review_id) DESC''' 

In [10]:
pd.io.sql.read_sql(query, con = engine)

title  count  round
0                              Twilight (Twilight  #1)      7    3.7
1                                  Water for Elephants      6    4.0
2                                     The Glass Castle      6    4.2
3    Harry Potter and the Prisoner of Azkaban (Harr...      6    4.4
4    The Curious Incident of the Dog in the Night-Time      6    4.1
..                                                 ...    ...    ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel      0    3.7
996                            The Natural Way to Draw      0    3.0
997   The Cat in the Hat and Other Dr. Seuss Favorites      0    5.0
998                          Essential Tales and Poems      0    4.0
999  Disney's Beauty and the Beast (A Little Golden...      0    4.0

[1000 rows x 3 columns]

Самое большое количество обзоров вышло на книгу - **Twilight (Twilight #1)** 

###  Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [11]:
query = ''' SELECT p.publisher,
                   COUNT(b.book_id)                   
            FROM books AS b 
            JOIN publishers AS p ON p.publisher_id = b.publisher_id  
            WHERE b.num_pages > 50 
            GROUP BY 1
            ORDER BY 2 DESC
            LIMIT 1'''

In [12]:
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

Издательство **Penguin Books** выпустило наибольшее число книг (42) толще 50 страниц

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [19]:
query = '''SELECT a.author, 
                  COUNT(rat.rating) AS count,
                  ROUND(AVG(rat.rating), 1) AS avg_rating
           FROM books AS b
           JOIN authors AS a ON a.author_id = b.author_id
           JOIN ratings AS rat ON b.book_id = rat.book_id
           WHERE b.book_id IN (SELECT book_id
                               FROM ratings 
                               GROUP BY book_id
                               HAVING COUNT (rating_id) >= 50
                               ORDER BY COUNT (rating_id) DESC,
                                        MAX(rating))
           
           GROUP BY 1
           ORDER BY 3 DESC
           LIMIT 1'''

In [20]:
pd.io.sql.read_sql(query, con = engine)

author  count  avg_rating
0  J.K. Rowling/Mary GrandPré    310         4.3

**J.K. Rowling/Mary GrandPré** - автор с самой высокой средней оценкой книг = 4.3 и количеством оценок = 312.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [15]:
query = ''' SELECT ROUND(AVG(two.count), 1) AS avg_review
            FROM
               (SELECT rev.username,
                        COUNT(rev.review_id) AS count
                 FROM       
                     (SELECT username,
                             COUNT(rating) 
                      FROM ratings
                      GROUP BY 1
                      HAVING COUNT(rating) > 50) AS one
                 JOIN reviews AS rev ON rev.username = one.username
                 GROUP BY 1) AS two
            '''

In [16]:
pd.io.sql.read_sql(query, con = engine)

avg_review
0        24.3

**24.3** - среднее количество обзоров от пользователей, совершивших больше 50 оценок.

## Вывод:
- **819** книг вышло после 1 января 2000 года
- Самое большое количество обзоров вышло на книгу - **Twilight (Twilight #1)** 
- Издательство **Penguin Books** выпустило наибольшее число книг (42) толще 50 страниц
- **J.K. Rowling/Mary GrandPré** - автор с самой высокой средней оценкой книг = 4.3 и количеством оценок = 312.
- **24.3** - среднее количество обзоров от пользователей, совершивших больше 50 оценок.